### PR TITLE
RUMM-979 PokoGenerator generate Type::fromJson deserializer method

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

A brief description of the change being made with this pull request.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: buildSrc/src/test/kotlin/com/example/model/Article.kt [hunk 1]
```diff
@@ -24,4 +31,42 @@ data class Article(
...
+                    val collection = ArrayList<String>(jsonArray.size())
+                    jsonArray.forEach {
+                        collection.add(it.asString)
+                    }
+                    collection
+                }
+                return Article(
+                title
+                ,tags
+                ,authors
```
Comment (xgouchet) [inline buildSrc/src/test/kotlin/com/example/model/Article.kt:61]: There's something wrong with the code style here.

**Review by xgouchet — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        val varName = property.name.variableName()
+        when (property.type) {
+            is TypeDefinition.Primitive -> appendPrimitiveDeserialization(
+                property,
+                property.type.type,
+                varName,
+                rootJsonObjectVarName,
+                funBuilder
+            )
+            is TypeDefinition.Null -> funBuilder.addStatement("val %L = null", property.name.trim())
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt:108]: it was acting weirdly without the trim, placing the variables with indents between them. I am going to try with the varName instead.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    }
+
+    /**
+     * Appends a primitive property deserialization to a [FunSpec.Builder].
+     * @param property the property definition
+     * @param type the primitive type
+     * @param rootJsonObjectVarName the root jsonObject var name
+     * @param funBuilder the `toJson()` [FunSpec] builder.
+     */
+    @Suppress("NON_EXHAUSTIVE_WHEN")
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt:138]: I think I just copy pasted this one from a previous one, I will remove that and add an JsonParse exception there.

File: buildSrc/src/test/kotlin/com/example/model/Style.kt [hunk 4]
```diff
@@ -13,6 +22,28 @@ data class Style(
...
+        @Throws(JsonParseException::class)
+        fun fromJson(serializedObject: String): Style {
+            try {
+                val jsonObject = JsonParser.parseString(serializedObject).asJsonObject
+                val color = jsonObject.get("color").asString.let {
+                    Color.valueOf(it.toUpperCase(Locale.US).replace(Regex("[^A-Z0-9]+"), "_"))
+                }
+                return Style(
+                color
+                )
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply buildSrc/src/test/kotlin/com/example/model/Style.kt:36]: Yeah.....not really sure how to fix that and didn't want to spend too much time on that for now. I will have another look.

File: buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+        } else {
+            funBuilder.beginControlFlow(
+                "val %L = %L.get(%S).asString.let",
+                varName.trim(),
+                rootJsonObjectVarName,
+                property.name
+            )
+        }
+        funBuilder.addStatement(
+            "%T.valueOf(it.toUpperCase(%T.US).replace(%T(\"[^A-Z0-9]+\"), \"_\"))",
```
[the remark this answers is not in the document]
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/PokoDeserializerGenerator.kt:380]: yeah...I don't like it either. What do you think about providing 2 extension methods in a separated file:
```fun String.toJsonValue()
     fun String.toEnumValue()
```

**Review by xgouchet — APPROVED** [review]

**Pull request merged**